### PR TITLE
fix bug with pod node affinity testing

### PR DESCRIPTION
**Pull request by tzneal**
**1. Issue, if available:**
N/A

**2. Description of changes:**

Adds a test and fixes a bug related to not checking pod node affinity.

**3. How was this change tested?**
Unit test + on EKS

**4. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [X] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:187]:
```suggestion
```

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:195]:
```suggestion
```

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:190]: Can't you just do 

```!it.OperatingSystems().HasAny(requirements.Get(v1.LabelOSStable))```

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:183]: Your node object already has `Constraints.Requirements`, and you can avoid the static.
```suggestion
func (n *Node) hasCompatibleInstanceType(it cloudprovider.InstanceType) bool {
```

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/node.go:183]: It's also used with a generated set of requirements from `hasCompatibleNodeAffinity`

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:195]:
```suggestion
	for _, offering := range it.Offerings() {
```

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:138]: Aren't these checks already modeled by `Requirements`? Does this disappear when we combine scheduling/binpacking? In the mean time, can you use PodRequirements(pod) to handle the selector and affinity parsing?

File: pkg/controllers/provisioning/scheduling/node.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/node.go:190]: I don't think so, the requirement may not specify an OS in which case any are ok.  We'd have to call .Values() on the requirements version which would panic due to calling values on a complement set.  The two different set types here are making the code more ugly than it should be.

File: pkg/controllers/provisioning/scheduling/node.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/node.go]: Yup 😞

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [reply pkg/controllers/provisioning/scheduling/node.go:119]: Another option is to implement HasAny on our specialized set type that encapsulates this. Totally optional.

File: pkg/controllers/provisioning/scheduling/node.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/node.go:61]: It feels like we're tripping over ourselves a bit. This is a tight loop that scales linearly with pods. We're already adding these requirements together inside `n.Constraints.Requirements.Compatible(podRequirements)`, and then we're doing it again here. Effectively, we have two areas of the code base responsible for the same thing (e.g. checking if a pod can schedule to a node), but done at different levels of abstraction.

As you've rightly done, we don't want to push this logic into the API package to create a dependency between `v1alpha5 -> cloudprovider`. Instead, if we could move the compatibility logic up to this layer, I *think* a lot of this impedance mismatch goes away. Obviously, we're blocked on this right now due to https://github.com/aws/karpenter/pull/1549, but it might be worth exploring once that's merged. 

I'm happy to merge this PR as it is currently, but would love to remove this rough edge for simplicity/performance.

File: pkg/controllers/provisioning/scheduling/node.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/node.go:119]: Good call, I didn't want to change much when fixing it for 0.7.2, but I've got the time now to make a bit less ugly.

File: pkg/controllers/provisioning/scheduling/node.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/node.go:60]: I was debating if the initial compatibility check was required when I wrote this and ended up with the conclusion that I wasn't certain that our Add() was written to handle possibly incompatible requirements, so it felt safer to keep it in.

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/node.go:60]: Just verified that the outer compatibility check handles the stuff that's not instance related (e.g. hostname requirements to implement topology spread).

Comment (ellistarn) [reply pkg/controllers/provisioning/scheduling/node.go:60]: Not currently. You'd need to smoosh em together :D

**Review by ellistarn — APPROVED** [review]

**Pull request merged**
